pkg/cortex/modules.go

```diff
@@ -190,55 +190,109 @@ func (t *Cortex) initDistributor() (serv services.Service, err error) {
+//           │Querier Queryable │  │  /api/v1 router  │  │ /api/prom router │
+//           │                  │  │                  │  │                  │
+//           └──────────────────┘  └──────────────────┘  └──────────────────┘
+//                     ▲                     │                     │
+//                     │                     └──────────┬──────────┘
+//                     │                                ▼
+//                     │                      ┌──────────────────┐
+//                     │                      │                  │
+//                     └──────────────────────│  Prometheus API  │
+//                                            │                  │
+//                                            └──────────────────┘
+func (t *Cortex) initQuerier() (serv services.Service, err error) {
+	// Create a internal HTTP handler that is configured with the Prometheus API routes and points
+	// to a Prometheus API struct instantiated with the Cortex Queryable.
+	internalQuerierRouter := api.NewQuerierHandler(
+		t.Cfg.API,
+		t.QuerierQueryable,
+		t.QuerierEngine,
+		t.Distributor,
+		t.TombstonesLoader,
+		prometheus.DefaultRegisterer,
+		util.Logger,
+	)
+
+	// If the querier is running standalone without the query-frontend, we must register it's internal
+	// HTTP handler externally and provide the external Cortex Server HTTP handler to the frontend worker
+	// to ensure requests it processes use the default middleware instrumentation.
+	if !t.Cfg.isModuleEnabled(QueryFrontend) && !t.Cfg.isModuleEnabled(All) {
+		// First, register the internal querier handler with the external HTTP server
+		t.API.RegisterQueryAPI(internalQuerierRouter)
+
+		// Second, set the http.Handler that the frontend worker will use to process requests to point to
+		// the external HTTP server. This will allow the querier to consolidate query metrics both external
+		// and internal using the default instrumentation when running as a standalone service.
+		internalQuerierRouter = t.Server.HTTPServer.Handler
+	} else {
+		// Single binary mode requires a query frontend endpoint for the worker. If no frontend endpoint
+		// is configured, Cortex will default to using localhost on it's own GRPC listening port.
+		if t.Cfg.Worker.Address == "" {
+			address := fmt.Sprintf("127.0.0.1:%d", t.Cfg.Server.GRPCListenPort)
+			level.Warn(util.Logger).Log("msg", "Worker address is empty in single binary mode.  Attempting automatic worker configuration.  If queries are unresponsive consider configuring the worker explicitly.", "address", address)
+			t.Cfg.Worker.Address = address
+		}
+
+		// If queries are processed using the external HTTP Server, we need wrap the internal querier with
+		// HTTP router with middleware to parse the tenant ID from the HTTP header and inject it into the
+		// request context.
+		internalQuerierRouter = middleware.AuthenticateUser.Wrap(internalQuerierRouter)
```

Comment:
We didn't do this before, did we? Why is it needed now?

Reply:
Actually, we did do this before because we previously used the `registerRouteWithRouter` function built into the API struct which applies this middleware. Now we just register the routes directly with with the router using the `Path|PathPrefix` function.